<a href="https://colab.research.google.com/github/mubarakdalvi/Data-Analytics-Projects/blob/main/Cointab_project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

importing libraries improtant for data i have imported matplot and seaborn as well but need to convert the data into numeric form to get perform some vizulization Didnt do EDA (Exploratory Data analysis)

In [722]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

rich library with many machine learning algorithms as i have learned them and happy to implement and this label encoder feature which is good

In [723]:
from sklearn.preprocessing import LabelEncoder

Data Loaded

In [724]:
from google.colab import files
data = files.upload()

Saving Company X - Order Report.xlsx to Company X - Order Report (3).xlsx
Saving Company X - Pincode Zones.xlsx to Company X - Pincode Zones (3).xlsx
Saving Company X - SKU Master.xlsx to Company X - SKU Master (3).xlsx
Saving Courier Company - Invoice.xlsx to Courier Company - Invoice (3).xlsx
Saving Courier Company - Rates.xlsx to Courier Company - Rates (3).xlsx
Saving Expected_Result.xlsx to Expected_Result (3).xlsx


checked for missing values in data and information about it

X company data which is right according to company

ETL Exract Transaform Load (ETL)

In [725]:
name1 = ['Order ID','Products','Order Qty']

contain orders in bundle like one order id contain alteast 5 to 6 items


In [726]:
CompanyXOrder_Report = pd.read_excel('Company X - Order Report.xlsx',names=name1)
CompanyXOrder_Report.head()

Order ID       Products  Order Qty
0  2001827036  8904223818706        1.0
1  2001827036  8904223819093        1.0
2  2001827036  8904223819109        1.0
3  2001827036  8904223818430        1.0
4  2001827036  8904223819277        1.0

In [727]:
CompanyXOrder_Report.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 400 entries, 0 to 399
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Order ID   400 non-null    int64  
 1   Products   400 non-null    object 
 2   Order Qty  400 non-null    float64
dtypes: float64(1), int64(1), object(1)
memory usage: 9.5+ KB


In [728]:
CompanyXOrder_Report.value_counts()

Order ID    Products       Order Qty
2001806885  8904223819499  2.0          2
2001811363  8904223815859  1.0          2
2001806210  8904223816214  1.0          1
2001811039  8904223818706  1.0          1
2001811192  8904223818874  1.0          1
                                       ..
2001807814  8904223818850  1.0          1
            8904223818706  1.0          1
2001807785  8904223818942  1.0          1
            8904223818850  1.0          1
2001827036  SACHETS001     1.0          1
Length: 398, dtype: int64

In [729]:
CompanyXOrder_Report['Order ID'].nunique()

124

In [730]:
CompanyXOrder_Report['Products'].nunique()

65

In [731]:
CompanyXOrder_Report.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 400 entries, 0 to 399
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Order ID   400 non-null    int64  
 1   Products   400 non-null    object 
 2   Order Qty  400 non-null    float64
dtypes: float64(1), int64(1), object(1)
memory usage: 9.5+ KB


from this can check which order i have have how many products in it

In [732]:
'''sortbyorder = CompanyXOrder_Report.sort_values(by='Order ID',ascending=False)'''

"sortbyorder = CompanyXOrder_Report.sort_values(by='Order ID',ascending=False)"

In [733]:
'''sortbyorder'''

'sortbyorder'

grouped_products will group orderid wise products so will get 124 entries which is good for next operation

In [734]:
'''grouped_products = CompanyXOrder_Report.groupby('Order ID')[['Products','Order Qty']].agg(list).reset_index()'''

"grouped_products = CompanyXOrder_Report.groupby('Order ID')[['Products','Order Qty']].agg(list).reset_index()"

In [735]:
'''grouped_products'''

'grouped_products'

In [736]:
CompanyXpincodes = ['Warehouse Pincode', 'Pincode by Company X', 'Zone By Comapany X']

In [737]:
CompanyXPincode_Zones = pd.read_excel('Company X - Pincode Zones.xlsx',names=CompanyXpincodes)
CompanyXPincode_Zones.head()

Warehouse Pincode  Pincode by Company X Zone By Comapany X
0             121003                507101                  d
1             121003                486886                  d
2             121003                532484                  d
3             121003                143001                  b
4             121003                515591                  d

In [738]:
CompanyXPincode_Zones.columns

Index(['Warehouse Pincode', 'Pincode by Company X', 'Zone By Comapany X'], dtype='object')

In [739]:
CompanyXPincode_Zones['Pincode by Company X'].nunique()

108

In [740]:
CompanyXPincode_Zones.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 124 entries, 0 to 123
Data columns (total 3 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   Warehouse Pincode     124 non-null    int64 
 1   Pincode by Company X  124 non-null    int64 
 2   Zone By Comapany X    124 non-null    object
dtypes: int64(2), object(1)
memory usage: 3.0+ KB


In [741]:
name2 = ['Products',	'Weight (g)']

In [742]:
CompanyXproduct_Master = pd.read_excel('Company X - SKU Master.xlsx',names=name2)
CompanyXproduct_Master.head()

Products  Weight (g)
0  8904223815682         210
1  8904223815859         165
2  8904223815866         113
3  8904223815873          65
4  8904223816214         120

weight conversion for better details

In [743]:
convert_weight = CompanyXproduct_Master['Weight (g)'] / 1000

adding in into CompanyXproduct_Master



In [744]:
CompanyXproduct_Master['Weight By Company X'] = convert_weight

renaming col weight to g to kg

In [745]:
CompanyXproduct_Master.rename_axis(columns={'Products' : 'Products','Weight (g)' : 'Weight (kg)'},inplace=True)

In [746]:
CompanyXproduct_Master.drop('Weight (g)',axis=1,inplace=True)

In [747]:
CompanyXproduct_Master

Products  Weight By Company X
0   8904223815682                0.210
1   8904223815859                0.165
2   8904223815866                0.113
3   8904223815873                0.065
4   8904223816214                0.120
..            ...                  ...
61  8904223819505                0.210
62  8904223819499                0.210
63  8904223819512                0.210
64  8904223819543                0.300
65     SACHETS001                0.010

[66 rows x 2 columns]

merging these two df to add weight into accoring to product

In [748]:
order_product = pd.merge(CompanyXOrder_Report,CompanyXproduct_Master)

In [749]:
order_product

Order ID       Products  Order Qty  Weight By Company X
0    2001827036  8904223818706        1.0                0.127
1    2001821995  8904223818706        1.0                0.127
2    2001819252  8904223818706        1.0                0.127
3    2001816996  8904223818706        1.0                0.127
4    2001814580  8904223818706        1.0                0.127
..          ...            ...        ...                  ...
396  2001806616  8904223819123        1.0                0.250
397  2001806567  8904223815804        1.0                0.160
398  2001806567  8904223818577        1.0                0.150
399  2001806408  8904223819437        2.0                0.552
400  2001806232  8904223818645        6.0                0.137

[401 rows x 4 columns]

grouping by order id to get all product into set



In [750]:
setofproduct = order_product.groupby('Order ID')['Products'].agg(list).reset_index()

In [751]:
setofproduct.head()

Order ID                                       Products
0  2001806210                 [8904223816214, 8904223818874]
1  2001806226                                [8904223818850]
2  2001806229  [8904223818706, 8904223818850, 8904223818942]
3  2001806232                 [8904223819147, 8904223818645]
4  2001806233                 [8904223819161, 8904223819260]

grouping

In [752]:
groupbysum = order_product.groupby('Order ID')[['Order Qty','Weight By Company X']].sum().reset_index()

In [753]:
combined_df = pd.merge(groupbysum, setofproduct, on='Order ID')
combined_df.head()

Order ID  Order Qty  Weight By Company X  \
0  2001806210        2.0                0.220   
1  2001806226        2.0                0.240   
2  2001806229        3.0                0.500   
3  2001806232        8.0                0.377   
4  2001806233        2.0                0.245   

                                        Products  
0                 [8904223816214, 8904223818874]  
1                                [8904223818850]  
2  [8904223818706, 8904223818850, 8904223818942]  
3                 [8904223819147, 8904223818645]  
4                 [8904223819161, 8904223819260]

In [754]:
combined_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 124 entries, 0 to 123
Data columns (total 4 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Order ID             124 non-null    int64  
 1   Order Qty            124 non-null    float64
 2   Weight By Company X  124 non-null    float64
 3   Products             124 non-null    object 
dtypes: float64(2), int64(1), object(1)
memory usage: 4.8+ KB


label encoder for getting values without if else as we didnt had any special criteria needed here so will use this

In [755]:
le = LabelEncoder()
Tran_OrderQty = le.fit_transform(groupbysum['Order Qty']) + 1

In [756]:
groupbysum['Order Qty'] = Tran_OrderQty

In [757]:
CompanyXproduct_Master.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 66 entries, 0 to 65
Data columns (total 2 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Products             66 non-null     object 
 1   Weight By Company X  66 non-null     float64
dtypes: float64(1), object(1)
memory usage: 1.2+ KB


This is Courier data Which is right or wrong will be decided by checking

In [758]:
courier_invoice = ['AWB Code', 'Order ID', 'Weight By Courier', 'Warehouse Pincode','Customer Pincode By Courier', 'Zone By Courier', 'Type of Shipment', 'Billing Amount (Rs.)']

In [759]:
Courier_Invoice  = pd.read_excel('Courier Company - Invoice.xlsx',names=courier_invoice)
Courier_Invoice.head()

AWB Code    Order ID  Weight By Courier  Warehouse Pincode  \
0  1091117222124  2001806232               1.30             121003   
1  1091117222194  2001806273               1.00             121003   
2  1091117222931  2001806408               2.50             121003   
3  1091117223244  2001806458               1.00             121003   
4  1091117229345  2001807012               0.15             121003   

   Customer Pincode By Courier Zone By Courier Type of Shipment  \
0                       507101               d  Forward charges   
1                       486886               d  Forward charges   
2                       532484               d  Forward charges   
3                       143001               b  Forward charges   
4                       515591               d  Forward charges   

   Billing Amount (Rs.)  
0                 135.0  
1                  90.2  
2                 224.6  
3                  61.3  
4                  45.4

In [760]:
Courier_Invoice.rename(columns={'Billing Amount (Rs.)':'Charges by Courier Company'},inplace=True)

In [761]:
Courier_Invoice.head()

AWB Code    Order ID  Weight By Courier  Warehouse Pincode  \
0  1091117222124  2001806232               1.30             121003   
1  1091117222194  2001806273               1.00             121003   
2  1091117222931  2001806408               2.50             121003   
3  1091117223244  2001806458               1.00             121003   
4  1091117229345  2001807012               0.15             121003   

   Customer Pincode By Courier Zone By Courier Type of Shipment  \
0                       507101               d  Forward charges   
1                       486886               d  Forward charges   
2                       532484               d  Forward charges   
3                       143001               b  Forward charges   
4                       515591               d  Forward charges   

   Charges by Courier Company  
0                       135.0  
1                        90.2  
2                       224.6  
3                        61.3  
4                        45.4

In [ ]:
grouping

In [762]:
grouobyinvoice = Courier_Invoice.groupby('Order ID')[['AWB Code','Weight By Courier','Zone By Courier','Charges by Courier Company','Customer Pincode By Courier']].agg(list).reset_index()

In [763]:
grouobyinvoice['Order ID']

0      2001806210
1      2001806226
2      2001806229
3      2001806232
4      2001806233
          ...    
119    2001821995
120    2001822466
121    2001823564
122    2001825261
123    2001827036
Name: Order ID, Length: 124, dtype: int64

In [764]:
Courier_Invoice['Zone By Courier'].unique()

array(['d', 'b', 'e'], dtype=object)

In [765]:
Courier_Invoice.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 124 entries, 0 to 123
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   AWB Code                     124 non-null    int64  
 1   Order ID                     124 non-null    int64  
 2   Weight By Courier            124 non-null    float64
 3   Warehouse Pincode            124 non-null    int64  
 4   Customer Pincode By Courier  124 non-null    int64  
 5   Zone By Courier              124 non-null    object 
 6   Type of Shipment             124 non-null    object 
 7   Charges by Courier Company   124 non-null    float64
dtypes: float64(2), int64(4), object(2)
memory usage: 7.9+ KB


In this Table we have Zones which is nothing but zones where customers are residing , as per zones weight slab is defined means 0.5kg is valid weight for
A zone  and have to pay if it have more than 0.5kg then

In [766]:
Courier_Rates = pd.read_excel('Courier Company - Rates.xlsx')
Courier_Rates.head()

Zone  Weight Slabs  Forward Fixed Charge  \
0    A          0.50                  29.5   
1    B          1.00                  33.0   
2    C          1.25                  40.1   
3    D          1.50                  45.4   
4    E          2.00                  56.6   

   Forward Additional Weight Slab Charge  RTO Fixed Charge  \
0                                   23.6              13.6   
1                                   28.3              20.5   
2                                   38.9              31.9   
3                                   44.8              41.3   
4                                   55.5              50.7   

   RTO Additional Weight Slab Charge  
0                               23.6  
1                               28.3  
2                               38.9  
3                               44.8  
4                               55.5

In [767]:
Courier_Rates.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 6 columns):
 #   Column                                 Non-Null Count  Dtype  
---  ------                                 --------------  -----  
 0   Zone                                   5 non-null      object 
 1   Weight Slabs                           5 non-null      float64
 2   Forward Fixed Charge                   5 non-null      float64
 3   Forward Additional Weight Slab Charge  5 non-null      float64
 4   RTO Fixed Charge                       5 non-null      float64
 5   RTO Additional Weight Slab Charge      5 non-null      float64
dtypes: float64(5), object(1)
memory usage: 368.0+ bytes


this is the result we want in last

In [903]:
'''Excpected_result = pd.read_excel('Expected_Result.xlsx')
Excpected_result.head()'''

"Excpected_result = pd.read_excel('Expected_Result.xlsx')\nExcpected_result.head()"

merged

In [769]:
order_awb = pd.merge(combined_df,grouobyinvoice,left_on='Order ID',right_on='Order ID')

In [770]:
order_awb.head()

Order ID  Order Qty  Weight By Company X  \
0  2001806210        2.0                0.220   
1  2001806226        2.0                0.240   
2  2001806229        3.0                0.500   
3  2001806232        8.0                0.377   
4  2001806233        2.0                0.245   

                                        Products         AWB Code  \
0                 [8904223816214, 8904223818874]  [1091117221940]   
1                                [8904223818850]  [1091117222065]   
2  [8904223818706, 8904223818850, 8904223818942]  [1091117222080]   
3                 [8904223819147, 8904223818645]  [1091117222124]   
4                 [8904223819161, 8904223819260]  [1091117222135]   

  Weight By Courier Zone By Courier Charges by Courier Company  \
0            [2.92]             [b]                    [174.5]   
1            [0.68]             [d]                     [90.2]   
2            [0.71]             [d]                     [90.2]   
3             [1.3]             [d]                    [135.0]   
4            [0.78]             [b]                     [61.3]   

  Customer Pincode By Courier  
0                    [140604]  
1                    [723146]  
2                    [421204]  
3                    [507101]  
4                    [263139]

In [771]:
order_awb['Order ID'] = order_awb['Order ID']

In [772]:
CompanyXPincode_Zones['Order ID'] = grouobyinvoice['Order ID']

In [773]:
CompanyXPincode_Zones

Warehouse Pincode  Pincode by Company X Zone By Comapany X    Order ID
0               121003                507101                  d  2001806210
1               121003                486886                  d  2001806226
2               121003                532484                  d  2001806229
3               121003                143001                  b  2001806232
4               121003                515591                  d  2001806233
..                 ...                   ...                ...         ...
119             121003                325207                  b  2001821995
120             121003                303702                  b  2001822466
121             121003                313301                  b  2001823564
122             121003                173212                  e  2001825261
123             121003                302020                  b  2001827036

[124 rows x 4 columns]

merge

In [941]:
zone_order = pd.merge(order_awb, CompanyXPincode_Zones, on='Order ID')

we have to remove square brakcet from them so we will user string astype and string to convert it to string and then we will need to convert int values back them to integers as we need in out operation

In [942]:
zone_order['Weight By Courier'] = zone_order['Weight By Courier'].astype(str).str.replace('[','').str.replace(']','')

<ipython-input-942-261f71243027>:1: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  zone_order['Weight By Courier'] = zone_order['Weight By Courier'].astype(str).str.replace('[','').str.replace(']','')


In [943]:
zone_order['Weight By Courier'] = pd.to_numeric(zone_order['Weight By Courier'])

In [944]:
zone_order['Products'] = zone_order['Products'].astype(str).str.replace('[','').str.replace(']',',')

<ipython-input-944-f3d9e31d6ede>:1: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  zone_order['Products'] = zone_order['Products'].astype(str).str.replace('[','').str.replace(']',',')


In [945]:
zone_order['Products'] = zone_order['Products'].str.rstrip(',')

In [946]:
zone_order['AWB Code'] = zone_order['AWB Code'].astype(str).str.replace('[','').str.replace(']','')

<ipython-input-946-3aefdd254934>:1: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  zone_order['AWB Code'] = zone_order['AWB Code'].astype(str).str.replace('[','').str.replace(']','')


In [947]:
zone_order['AWB Code'] = pd.to_numeric(zone_order['AWB Code'])

In [948]:
zone_order['Zone By Courier'] = zone_order['Zone By Courier'].astype(str).str.replace('[','').str.replace(']','')

<ipython-input-948-b1203af8ede6>:1: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  zone_order['Zone By Courier'] = zone_order['Zone By Courier'].astype(str).str.replace('[','').str.replace(']','')


In [949]:
zone_order['Zone By Courier'] = zone_order['Zone By Courier'].astype(str).str.replace('[','').str.replace(']','')

<ipython-input-949-b1203af8ede6>:1: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  zone_order['Zone By Courier'] = zone_order['Zone By Courier'].astype(str).str.replace('[','').str.replace(']','')


In [950]:
zone_order['Charges by Courier Company'] = zone_order['Charges by Courier Company'].astype(str).str.replace('[','').str.replace(']','')

<ipython-input-950-1a4923cfda33>:1: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  zone_order['Charges by Courier Company'] = zone_order['Charges by Courier Company'].astype(str).str.replace('[','').str.replace(']','')


In [951]:
zone_order['Charges by Courier Company'] = pd.to_numeric(zone_order['Charges by Courier Company'])

In [952]:
zone_order['Zone By Courier'] = zone_order['Zone By Courier'].astype(object).str.replace('[','').str.replace(']','')

<ipython-input-952-469fd2469c7c>:1: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  zone_order['Zone By Courier'] = zone_order['Zone By Courier'].astype(object).str.replace('[','').str.replace(']','')


as for here we are doing it for replacing singel qoutes from zone by couriers

In [953]:
zone_order['Zone By Courier'] = zone_order['Zone By Courier'].astype(object).str.replace("'",'').str.replace("'",'')

In [954]:
zone_order['Customer Pincode By Courier'] = zone_order['Customer Pincode By Courier'].astype(str).str.replace('[','').str.replace(']','')

<ipython-input-954-1b2e72fd8f34>:1: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  zone_order['Customer Pincode By Courier'] = zone_order['Customer Pincode By Courier'].astype(str).str.replace('[','').str.replace(']','')


this will convert to numeric values

In [955]:
zone_order['Customer Pincode By Courier'] = pd.to_numeric(zone_order['Customer Pincode By Courier'])

In [956]:
zone_order.head()

Order ID  Order Qty  Weight By Company X  \
0  2001806210        2.0                0.220   
1  2001806226        2.0                0.240   
2  2001806229        3.0                0.500   
3  2001806232        8.0                0.377   
4  2001806233        2.0                0.245   

                                      Products       AWB Code  \
0                 8904223816214, 8904223818874  1091117221940   
1                                8904223818850  1091117222065   
2  8904223818706, 8904223818850, 8904223818942  1091117222080   
3                 8904223819147, 8904223818645  1091117222124   
4                 8904223819161, 8904223819260  1091117222135   

   Weight By Courier Zone By Courier  Charges by Courier Company  \
0               2.92               b                       174.5   
1               0.68               d                        90.2   
2               0.71               d                        90.2   
3               1.30               d                       135.0   
4               0.78               b                        61.3   

   Customer Pincode By Courier  Warehouse Pincode  Pincode by Company X  \
0                       140604             121003                507101   
1                       723146             121003                486886   
2                       421204             121003                532484   
3                       507101             121003                143001   
4                       263139             121003                515591   

  Zone By Comapany X  
0                  d  
1                  d  
2                  d  
3                  b  
4                  d

alternattive methods. for string conversion.

In [957]:
'''df['charge'] = df['charge'].apply(lambda x: x[0])'''

"df['charge'] = df['charge'].apply(lambda x: x[0])"

getting values as per zones for company X


In [958]:
def impute_row(row):
  if row['Zone By Comapany X'] == 'a':
    return 0.50
  elif row['Zone By Comapany X'] == 'b':
    return 1.00
  elif row['Zone By Comapany X'] == 'c':
    return 1.25
  elif row['Zone By Comapany X'] == 'd':
    return 1.50
  elif row['Zone By Comapany X'] == 'e':
    return 2.00
zone_order['Weight slab (kg)'] = zone_order.apply(impute_row,axis=1)

getting values as per zones for courier company

In [959]:
def wscourier(row):
  if row['Zone By Courier'] == 'a':
    return 0.50
  elif row['Zone By Courier'] == 'b':
    return 1.00
  elif row['Zone By Courier'] == 'c':
    return 1.25
  elif row['Zone By Courier'] == 'd':
    return 1.50
  elif row['Zone By Courier'] == 'e':
    return 2.00

zone_order['weight slab by Courier'] = zone_order.apply(wscourier,axis=1)

In [960]:
zone_order.head()

Order ID  Order Qty  Weight By Company X  \
0  2001806210        2.0                0.220   
1  2001806226        2.0                0.240   
2  2001806229        3.0                0.500   
3  2001806232        8.0                0.377   
4  2001806233        2.0                0.245   

                                      Products       AWB Code  \
0                 8904223816214, 8904223818874  1091117221940   
1                                8904223818850  1091117222065   
2  8904223818706, 8904223818850, 8904223818942  1091117222080   
3                 8904223819147, 8904223818645  1091117222124   
4                 8904223819161, 8904223819260  1091117222135   

   Weight By Courier Zone By Courier  Charges by Courier Company  \
0               2.92               b                       174.5   
1               0.68               d                        90.2   
2               0.71               d                        90.2   
3               1.30               d                       135.0   
4               0.78               b                        61.3   

   Customer Pincode By Courier  Warehouse Pincode  Pincode by Company X  \
0                       140604             121003                507101   
1                       723146             121003                486886   
2                       421204             121003                532484   
3                       507101             121003                143001   
4                       263139             121003                515591   

  Zone By Comapany X  Weight slab (kg)  weight slab by Courier  
0                  d               1.5                     1.0  
1                  d               1.5                     1.5  
2                  d               1.5                     1.5  
3                  b               1.0                     1.5  
4                  d               1.5                     1.0

calculating charges as per in simple we will calculate on the basis of weight and slab by zones and if weight is more than zone we will add addional charges in case of rto will add addtional rto charges as fixed rto charges will be add , because it is fixed charges

In [961]:
def calculate_charge(row):
    Zone = ['a', 'b', 'c', 'd', 'e']
    WeightSlabs = [0.50, 1.00, 1.25, 1.50, 2.00]
    ForwardFixedCharge = [29.5, 33.0, 40.1, 45.4, 56.6]
    ForwardAdditionalWeightSlabCharge = [23.6, 28.3, 38.9, 44.8, 55.5]
    RTOFixedCharge = [13.6, 20.5, 31.9, 41.3, 50.7]
    RTOAdditionalWeightSlabCharge = [23.6, 28.3, 38.9, 44.8, 55.5]

    zone_index = Zone.index(row['Zone By Comapany X'])
    fixed_weight_slab = WeightSlabs[zone_index]
    remaining_weight = max(0, row['Weight By Company X'] - fixed_weight_slab)

    charge = ForwardFixedCharge[zone_index]
    if remaining_weight > 0:
        charge += remaining_weight * ForwardAdditionalWeightSlabCharge[zone_index]

    if row['Weight By Company X']:
        charge += RTOFixedCharge[zone_index]
        if remaining_weight > 0:
            charge += remaining_weight * RTOAdditionalWeightSlabCharge[zone_index]

    return charge

zone_order['Expected By Company X'] = zone_order.apply(calculate_charge, axis=1)

In [962]:
zone_order.head()

Order ID  Order Qty  Weight By Company X  \
0  2001806210        2.0                0.220   
1  2001806226        2.0                0.240   
2  2001806229        3.0                0.500   
3  2001806232        8.0                0.377   
4  2001806233        2.0                0.245   

                                      Products       AWB Code  \
0                 8904223816214, 8904223818874  1091117221940   
1                                8904223818850  1091117222065   
2  8904223818706, 8904223818850, 8904223818942  1091117222080   
3                 8904223819147, 8904223818645  1091117222124   
4                 8904223819161, 8904223819260  1091117222135   

   Weight By Courier Zone By Courier  Charges by Courier Company  \
0               2.92               b                       174.5   
1               0.68               d                        90.2   
2               0.71               d                        90.2   
3               1.30               d                       135.0   
4               0.78               b                        61.3   

   Customer Pincode By Courier  Warehouse Pincode  Pincode by Company X  \
0                       140604             121003                507101   
1                       723146             121003                486886   
2                       421204             121003                532484   
3                       507101             121003                143001   
4                       263139             121003                515591   

  Zone By Comapany X  Weight slab (kg)  weight slab by Courier  \
0                  d               1.5                     1.0   
1                  d               1.5                     1.5   
2                  d               1.5                     1.5   
3                  b               1.0                     1.5   
4                  d               1.5                     1.0   

   Expected By Company X  
0                   86.7  
1                   86.7  
2                   86.7  
3                   53.5  
4                   86.7

getting difference between courier and x company charges

In [963]:
zone_order['Difference Between Charges'] = zone_order['Charges by Courier Company'] - zone_order['Expected By Company X']

In [964]:
zone_order.head()

Order ID  Order Qty  Weight By Company X  \
0  2001806210        2.0                0.220   
1  2001806226        2.0                0.240   
2  2001806229        3.0                0.500   
3  2001806232        8.0                0.377   
4  2001806233        2.0                0.245   

                                      Products       AWB Code  \
0                 8904223816214, 8904223818874  1091117221940   
1                                8904223818850  1091117222065   
2  8904223818706, 8904223818850, 8904223818942  1091117222080   
3                 8904223819147, 8904223818645  1091117222124   
4                 8904223819161, 8904223819260  1091117222135   

   Weight By Courier Zone By Courier  Charges by Courier Company  \
0               2.92               b                       174.5   
1               0.68               d                        90.2   
2               0.71               d                        90.2   
3               1.30               d                       135.0   
4               0.78               b                        61.3   

   Customer Pincode By Courier  Warehouse Pincode  Pincode by Company X  \
0                       140604             121003                507101   
1                       723146             121003                486886   
2                       421204             121003                532484   
3                       507101             121003                143001   
4                       263139             121003                515591   

  Zone By Comapany X  Weight slab (kg)  weight slab by Courier  \
0                  d               1.5                     1.0   
1                  d               1.5                     1.5   
2                  d               1.5                     1.5   
3                  b               1.0                     1.5   
4                  d               1.5                     1.0   

   Expected By Company X  Difference Between Charges  
0                   86.7                        87.8  
1                   86.7                         3.5  
2                   86.7                         3.5  
3                   53.5                        81.5  
4                   86.7                       -25.4

getting in output1 as per desired columns

In [965]:
output1 = zone_order[['Order ID','AWB Code','Weight By Company X','Weight slab (kg)','Weight By Courier','weight slab by Courier','Zone By Comapany X','Zone By Courier','Expected By Company X','Charges by Courier Company','Difference Between Charges']]

In [967]:
output1.head()

Order ID       AWB Code  Weight By Company X  Weight slab (kg)  \
0  2001806210  1091117221940                0.220               1.5   
1  2001806226  1091117222065                0.240               1.5   
2  2001806229  1091117222080                0.500               1.5   
3  2001806232  1091117222124                0.377               1.0   
4  2001806233  1091117222135                0.245               1.5   

   Weight By Courier  weight slab by Courier Zone By Comapany X  \
0               2.92                     1.0                  d   
1               0.68                     1.5                  d   
2               0.71                     1.5                  d   
3               1.30                     1.5                  b   
4               0.78                     1.0                  d   

  Zone By Courier  Expected By Company X  Charges by Courier Company  \
0               b                   86.7                       174.5   
1               d                   86.7                        90.2   
2               d                   86.7                        90.2   
3               d                   53.5                       135.0   
4               b                   86.7                        61.3   

   Difference Between Charges  
0                        87.8  
1                         3.5  
2                         3.5  
3                        81.5  
4                       -25.4

data frame to excel convert

In [968]:
output1.to_excel('output1.xlsx')

google colab feature

In [ ]:
'''from google.colab import files
data = files.download('output1.xlsx')'''

In [ ]:
zone_order.columns

for getting every other details like count and sum of the expected and x courier charges

In [889]:
zone_order['Charging Correct'] = zone_order['Charges by Courier Company'] == zone_order['Expected By Company X']

total_correct_count = zone_order['Charging Correct'].sum()
total_overcharged_count = (zone_order['Charges by Courier Company'] > zone_order['Expected By Company X']).sum()
total_undercharged_count = (zone_order['Charges by Courier Company'] < zone_order['Expected By Company X']).sum()

total_correct_amount = zone_order.loc[zone_order['Charging Correct'], 'Charges by Courier Company'].sum()
total_overcharged_amount = zone_order.loc[zone_order['Charges by Courier Company'] > zone_order['Expected By Company X'], 'Charges by Courier Company'].sum()
total_undercharged_amount = zone_order.loc[zone_order['Charges by Courier Company'] < zone_order['Expected By Company X'], 'Charges by Courier Company'].sum()


applying to new column named with them

In [890]:
zone_order['Count_Charging_Correct'] = total_correct_count
zone_order['total_overcharged_count'] = total_overcharged_count
zone_order['total_undercharged_count'] = total_undercharged_count
zone_order['total_correct_amount'] = total_correct_amount
zone_order['total_overcharged_amount'] = total_overcharged_amount
zone_order['total_undercharged_amount'] = total_undercharged_amount

preparin for ouput 2

In [891]:
check = zone_order[['Count_Charging_Correct','total_overcharged_count','total_undercharged_count','total_correct_amount','total_overcharged_amount','total_undercharged_amount']]

In [ ]:
check

In [898]:
output2 = check.iloc[0]

In [ ]:
output2

In [901]:
output2.to_excel('output2.xlsx')

downlod with this feature

In [902]:
'''from google.colab import files
data = files.download('output2.xlsx')'''

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>